enterprise/server/backends/migration_cache/migration_cache.go

```diff
@@ -146,31 +148,174 @@ func pebbleCacheFromConfig(env environment.Env, cfg *PebbleCacheConfig) (*pebble
+	srcCache, err := mc.src.WithIsolation(ctx, cacheType, remoteInstanceName)
+	if err != nil {
+		return nil, errors.WithMessage(err, "cannot get src cache with isolation")
+	}
+	destCache, err := mc.dest.WithIsolation(ctx, cacheType, remoteInstanceName)
+	if err != nil {
+		return nil, errors.WithMessage(err, "cannot get dest cache with isolation")
+	}
+
+	clone := *mc
+	clone.src = srcCache
+	clone.dest = destCache
+
+	return &clone, nil
 }
 
 func (mc *MigrationCache) Contains(ctx context.Context, d *repb.Digest) (bool, error) {
-	return false, status.UnimplementedError("not yet implemented")
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+	var srcContains, dstContains bool
+
+	eg.Go(func() error {
+		srcContains, srcErr = mc.src.Contains(gctx, d)
+		return srcErr
+	})
+
+	doubleRead := rand.Float64() <= mc.doubleReadPercentage
+	if doubleRead {
+		eg.Go(func() error {
+			dstContains, dstErr = mc.dest.Contains(gctx, d)
+			return nil // we don't care about the return error from this cache
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return false, err
+	}
+
+	if dstErr != nil {
+		log.Warningf("Migration dest %v contains failed: %s", d, dstErr)
+	} else if mc.logNotFoundErrors && srcContains != dstContains {
+		log.Warningf("Migration digest %v src contains %v, dest contains %v", d, srcContains, dstContains)
+	}
+
+	return srcContains, srcErr
 }
 
 func (mc *MigrationCache) Metadata(ctx context.Context, d *repb.Digest) (*interfaces.CacheMetadata, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+	var srcMetadata *interfaces.CacheMetadata
+
+	eg.Go(func() error {
+		srcMetadata, srcErr = mc.src.Metadata(gctx, d)
+		return srcErr
+	})
+
+	doubleRead := rand.Float64() <= mc.doubleReadPercentage
+	if doubleRead {
+		eg.Go(func() error {
+			_, dstErr = mc.dest.Metadata(gctx, d)
+			return nil // we don't care about the return error from this cache
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	if dstErr != nil && (mc.logNotFoundErrors || !status.IsNotFoundError(dstErr)) {
+		log.Warningf("Migration dest %v metadata failed: %s", d, dstErr)
+	}
+
+	return srcMetadata, srcErr
 }
 
 func (mc *MigrationCache) FindMissing(ctx context.Context, digests []*repb.Digest) ([]*repb.Digest, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+	var srcMissing []*repb.Digest
+
+	eg.Go(func() error {
+		srcMissing, srcErr = mc.src.FindMissing(gctx, digests)
+		return srcErr
+	})
+
+	doubleRead := rand.Float64() <= mc.doubleReadPercentage
+	if doubleRead {
+		eg.Go(func() error {
+			_, dstErr = mc.dest.FindMissing(gctx, digests)
+			return nil // we don't care about the return error from this cache
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	if dstErr != nil {
```

Comment:
`FindMissing` is a core API for Bazel caching. We should verify that the two caches are returning consistent results.

Reply:
Done